### ISSUE TITLE
RemoveUnusedImports not working

### ISSUE THREAD

**Issue by avdv** [scalacenter/scalafix#592]
Hi.

I am trying to use scalafix 0.5.10 with a SBT project. A quick test with `scalafix ProcedureSyntax` was successful.

But running `scalafix RemoveUnusedImports` does not change anything, although the Scala compiler spits out 11 "Unused import" warnings:

```
[warn] /home/claudio/src/sbt-eve-play/src/main/scala/com/redbull/eve/sbt/EveCommonPlugin.scala:81:5: Unused import
[warn]     commandLineReleaseVersion,  useDefaults, versions
[warn]     ^
[warn] /home/claudio/src/sbt-eve-play/src/main/scala/com/redbull/eve/sbt/EveCommonPlugin.scala:81:33: Unused import
[warn]     commandLineReleaseVersion,  useDefaults, versions
[warn]                                 ^
[warn] /home/claudio/src/sbt-eve-play/src/main/scala/com/redbull/eve/sbt/EveCommonPlugin.scala:81:46: Unused import
[warn]     commandLineReleaseVersion,  useDefaults, versions
...
```
The imports are still there.

I also tried to run scalafix manually, using the `--stdout` option, but with the same result. The files are printed as they are processed, but still contain the unused imports.

Using `--verbose` does not help either. What can I do to debug this?

**Comment by MasseGuillaume** [comment]
Do you have scalafix enabled in your "*build build*"?

`project/project/plugins.sbt` must contain `addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.5.10")`

**Comment by MasseGuillaume** [comment]
Hmm nevermind, I thought those warnings are from the build itself, they are from an sbt plugin.

**Comment by olafurpg** [comment]
@avdv can you provide a minimal reproduction? Ideally from the standard library. Is this a regression from 0.5.7? What happens if you run `scalafixCli --rules RemoveUnusedImports`?

**Comment by avdv** [comment]
Sorry, should have been more specific. Yes, this is from an SBT plugin, but I also tried it in another SBT project.

I am using SBT 1.1.0, Scala 2.12.4. The scalacOptions are looking OK and the semanticdb files are generated.

**Comment by olafurpg** [comment]
What happens if you run scalafixCli --rules RemoveUnusedImports?

**Comment by olafurpg** [comment]
Is this a regression from 0.5.7?

**Comment by olafurpg** [comment]
Can you provide a small source file that reproduces the bug? :)

**Comment by avdv** [comment]
Hi @olafurpg,

I have tried with 0.5.7, but still no dice. I have prepared a fresh SBT project with `sbt new scala/hello-world.g8`, set the Scala version to 2.12.4, SBT version to 1.1.0 and added the scalafix plugin. It did not work.

Also, I had used scalafix downloaded with coursier manually to no avail.

But then, I ran SBT with `-sbt-dir /tmp/sbt` and low and behold, this time it worked. Seems like scalafix conflicts with some global SBT plugins (namely coursier?!) I have activated... I'll look into this.

**Comment by avdv** [comment]
OK, seems the culprit is not coursier as suspected, but [clippy](https://scala-clippy.org/):
```
addSbtPlugin("com.softwaremill.clippy" % "plugin-sbt" % "0.5.3")
```
I still do not understand why running scalafix manually did not work either. (I just ran `~/.coursier/coursier bootstrap ch.epfl.scala:scalafix-cli_2.12.4:0.5.10 -f --main scalafix.cli.Cli -o scalafix` to create an executable). Does coursier use SBT under the hood - ie. also the global plugins?

**Comment by olafurpg** [comment]
Thanks for investigating @avdv Interesting, I suspect there is a bad interplay between the clippy compiler plugin and semanticdb-scalac. RemoveUnusedImports uses the messages reported by the compiler so this bug makes sense if clippy is changing the reporter

**Comment by olafurpg** [comment]
clippy looks btw super cool! Want to try it out 😄

**Comment by ShaneDelmore** [comment]
Clippy does hijack the reporter and inject new messages but it also leaves the original message so it should be possible to make them play well together.

**Comment by avdv** [comment]
thanks @ShaneDelmore, that makes total sense now, I compiled (read: generated the semanticdb's) with clippy **on**, so running scalafix on the command line later did still not recognize the import errors -- same input -> same output.

> clippy looks btw super cool! Want to try it out 😄

Oh yes, I guess I'm so used to its output that I forgot that the messages look different usually... 😉

**Comment by avdv** [comment]
There is not much that could be done by scalafix about this, I think. Should we close this issue?

**Comment by MasseGuillaume** [comment]
@avdv we should add a note about this in the documentation. Maybe open an issue on the clippy side. I'm not sure why we are not being able to consume those errors.

**Comment by avdv** [comment]
Yes, I did open an issue here: https://github.com/softwaremill/scala-clippy/issues/61

Somehow clippy does not leave the original message in the semanticdb file.

**Comment by olafurpg** [comment]
The relevant piece of code in scalameta is https://github.com/scalameta/scalameta/blob/2bc7cfd081b45bfbb2e16c60eb906fa208a597cb/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/HijackReporter.scala#L12 We set the reporter to our own custom implementation that records all messages and forwards them to the underlying reporter.

The ordering of the compiler plugins may impact if it works or not. You can try `scalafixEnabled := false` and manually `addCompilerPlugin("org.scalameta" % "semanticdb-scalac" ....)`. Not sure whether it's better to enable semanticdb before or after clippy.
